contracts/Dripper.sol

```diff
@@ -272,4 +254,129 @@ contract Dripper is Ownable {
+        uint256 price = _getConversionPrice();
+        uint256 twap = _getConversionTWAP();
+        require(
+            twap.mul(ONE).div(price) < ONE.add(dripConfig.maxTWAPDifferencePct)
+            && price.mul(ONE).div(twap) < ONE.add(dripConfig.maxTWAPDifferencePct),
+            ERROR_FAR_FROM_TWAP
+        );
+        return price;
+    }
+
+    /**
+     * @notice Calculates how much startLP tokens we have to drip right now
+     */
+    function _calculateDrip() internal view returns (uint256 dripAmt) {
+        // calculate how much should be withdrawn
+        uint256 timeSinceLastDrip = block.timestamp.sub(
+            latestDripTime < dripConfig.startTime ?
+                dripConfig.startTime : latestDripTime
+        );
+        uint256 startLPToWithdraw = dripConfig.amountToDrip.mul(
+                timeSinceLastDrip.mul(ONE).div(dripConfig.transitionTime)
+            ).div(ONE);
+
+        return startLPToWithdraw;
+    }
+
+    /**
+     * @notice Swaps `amountIn` start tokens for end tokens
+     * @param amountIn the amount of startTokens to swap
+     * @return amountOut the output of the swap in endTokens
+     */
+    function _swapTokens(uint256 amountIn) internal returns (uint256 amountOut) {
+        // swap fromToken to endToken
+        address[] memory path = new address[](2);
+        path[0] = address(startToken);
+        path[1] = address(endToken);
+
+        // Calculate minimum expected output, taking slippage into account
+        uint256 expectedOutput = _getQuote(address(startToken), address(endToken)).mul(amountIn).div(ONE);
+        expectedOutput = expectedOutput.mul(ONE.sub(dripConfig.maxSlippageTolerancePct)).div(ONE);
+
+        startToken.approve(address(router), amountIn);
+
+        uint[] memory amounts = router.swapExactTokensForTokens(
+            amountIn,
+            expectedOutput,
+            path,
+            address(this),
+            now + 1
+        );
+
+        return amounts[amounts.length.sub(1)];
+    }
+
+    /**
+     * @notice Optimize token amounts for addition into a liquidity pool
+     * @param endAmount initial guess on how much endTokens to add
+     * @param baseAmount initial guess on how much baseTokens to add
+     * @return optimized endTokenAmount and baseTokenAmount
+     */
+    function _optimizeAmounts(uint256 endAmount, uint256 baseAmount) internal view returns (uint256, uint256) {
+        // optimize token amounts to add as liquidity
+        // based on the amount tokens we've got
+        uint256 quote = _getQuote(address(endToken), address(baseToken));
+        uint256 optimizedBaseTokenAmount = quote.mul(endAmount).div(ONE);
+        uint256 optimizedEndTokenAmount = ONE.div(quote).mul(baseAmount);
+
+        // optimize one side, if not, optimize the other.
+        if (optimizedBaseTokenAmount <= baseAmount){
+            return (endAmount, optimizedBaseTokenAmount);
+        } else if (optimizedEndTokenAmount <= endAmount){
```

Comment:
This should probably be an else. Otherwise this could return with 0, 0 unexpectedly.

Reply:
Fixed in latest diff.